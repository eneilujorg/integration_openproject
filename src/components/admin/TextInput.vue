<template>
	<div class="text-input-wrapper">
		<div class="text-input">
			<div class="text-input-label">
				{{ translate(labelText) }}
			</div>
			<input :id="id"
				ref="textInput"
				:value="value"
				:type="type"
				class="full-width"
				:class="{'error': !!errorMessage}"
				:placeholder="translate(placeHolder)"
				@input="$emit('input', $event.target.value)"
				@change="$emit('change', $event.target.value)"
				@focus="$emit('focus', $event)"
				@blur="$emit('blur', $event)">
			<div v-if="errorMessage || hintText" class="text-input-messages">
				<div v-if="errorMessage" class="text-input-error">
					{{ translate(errorMessage) }}
				</div>
				<div v-else
					class="text-input-hint"
					v-html="hintText" />
			</div>
		</div>
		<button v-if="withCopyBtn"
			class="copy-btn"
			:disabled="isCopyDisabled"
			@click="copyValue">
			<div class="copy-icon" />
			<span>{{ translate("Copy value") }}</span>
		</button>
	</div>
</template>
<script>
import { translate as t } from '@nextcloud/l10n'

export default {
	name: 'TextInput',
	props: {
		value: {
			default: '',
			type: String,
		},
		id: {
			type: String,
			required: true,
		},
		type: {
			type: String,
			default: 'text',
		},
		placeHolder: {
			default: '',
			type: String,
		},
		label: {
			required: true,
			type: String,
		},
		hintText: {
			default: null,
			type: [String, null],
		},
		errorMessage: {
			default: null,
			type: [String, null],
		},
		isRequired: {
			default: false,
			type: Boolean,
		},
		withCopyBtn: {
			default: false,
			type: Boolean,
		},
	},
	computed: {
		labelText() {
			if (this.isRequired) {
				return `${this.label} *`
			} else return this.label
		},
		isCopyDisabled() {
			return !this.value
		},
	},
	methods: {
		translate(text) {
			return t('integration_openproject', text)
		},
		copyValue() {
			navigator.clipboard.writeText(this.value)
		},
	},
}
</script>
<style lang="scss" scoped>
.text-input-wrapper {
	display: flex;
	align-items: center;
}

.text-input {
	.full-width {
		width: 100%;
	}

	&-label {
		font-weight: 700;
		font-size: .875rem;
		line-height: 1.25rem;
		color: #333333;
	}

	&-messages {
		font-weight: 400;
		font-size: .75rem;
		line-height: 1rem;
	}

	.error {
		border: 2px solid var(--color-error) !important;
	}

	&-error {
		color: var(--color-error);
	}
	input[data-focus-visible-added].error {
		outline: none;
		box-shadow: none;
	}
}

.copy-btn {
	cursor: copy;
	display: flex;
	align-items: center;
	margin-left: 10px;
	margin-top: 6px;
	span {
		cursor: copy;
		margin-left: 6px;
	}
	.copy-icon {
		cursor: copy;
		width: 16px;
		height: 16px;
		background-size: 16px;
		background-repeat: no-repeat;
		background-position: center;
		background-image: url(./../../../img/copy.svg);
	}
}

body[data-theme-dark-highcontrast], body[data-theme-dark], body.theme--dark {
	.text-input {
		&-label {
			color: #FFFFFF;
		}
	}
}
</style>
<style>
.text-input .link {
	color: #1a67a3 !important;
	font-style: italic;
}
</style>
