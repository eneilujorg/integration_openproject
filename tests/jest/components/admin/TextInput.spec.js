import { createLocalVue, mount } from '@vue/test-utils'
import TextInput from '../../../../src/components/admin/TextInput'

const localVue = createLocalVue()

Object.assign(navigator, {
	clipboard: {
		writeText: () => {},
	},
})

jest.mock('@nextcloud/dialogs', () => ({
	showSuccess: jest.fn(),
}))

const selector = {
	textInputLabel: '.text-input-label',
	copyButton: '.text-input-copy-value',
	copyIcon: '.icon-clippy',
}

describe('TextInput', () => {
	describe('messages', () => {
		it('should show error message if provided', () => {
			const wrapper = getWrapper({
				hintText: null,
				errorMessage: 'some error message',
			})
			expect(wrapper).toMatchSnapshot()
		})
		it('should show hint text if provided', () => {
			const wrapper = getWrapper({
				hintText: 'some hint message',
				errorMessage: null,
			})
			expect(wrapper).toMatchSnapshot()
		})
		it('should show error message if both error message and hint text are provided', () => {
			const wrapper = getWrapper({
				hintText: 'some hint message',
				errorMessage: 'some error message',
			})
			expect(wrapper).toMatchSnapshot()
		})
	})
	describe('is required prop', () => {
		it('should add asterik to the label text', () => {
			const wrapper = getWrapper({
				isRequired: true,
			})
			expect(wrapper.find(selector.textInputLabel)).toMatchSnapshot()
		})
		it('should not add asterik to the label text', () => {
			const wrapper = getWrapper({
				isRequired: false,
			})
			expect(wrapper.find(selector.textInputLabel)).toMatchSnapshot()
		})
	})
	describe('with copy button prop', () => {
		let wrapper
		beforeEach(() => {
			wrapper = getWrapper({
				withCopyBtn: true,
			})
		})
		it('should render copy button if set', () => {
			expect(wrapper).toMatchSnapshot()
		})
		it('should be disabled if the input value is empty', () => {
			expect(wrapper.find(selector.copyButton).attributes().disabled).toBe('disabled')
		})
		it('should be enabled if the input value is non empty', async () => {
			wrapper = getWrapper({
				withCopyBtn: true,
				value: 'some-value',
			})
			expect(wrapper.find(selector.copyButton).attributes().disabled).toBe(undefined)
		})
		describe('on click', () => {
			let copyButton
			jest.useFakeTimers()
			const spyWriteToClipboard = jest.spyOn(navigator.clipboard, 'writeText')
				.mockImplementationOnce(() => jest.fn())
			beforeEach(() => {
				wrapper = getWrapper({
					withCopyBtn: true,
					value: 'some-value-to-copy',
				})
				copyButton = wrapper.find(selector.copyButton)
			})
			it('should copy the input value', async () => {
				await copyButton.trigger('click')
				expect(spyWriteToClipboard).toBeCalledTimes(1)
				expect(spyWriteToClipboard).toBeCalledWith('some-value-to-copy')
			})
			it('should change the copy icon with the copied icon', async () => {
				expect(copyButton.attributes().title).toBe('Copy value')
				await copyButton.trigger('click')
				await wrapper.vm.$nextTick()
				copyButton = wrapper.find(selector.copyButton)
				expect(copyButton.attributes().title).toBe('Copied!')
				jest.advanceTimersByTime(5000)
				await wrapper.vm.$nextTick()
				copyButton = wrapper.find(selector.copyButton)
				expect(copyButton.attributes().title).toBe('Copy value')
			})
		})
	})
})

function getWrapper(props = {}) {
	return mount(TextInput, {
		localVue,
		propsData: {
			value: null,
			id: 'unique-id',
			label: 'some label',
			...props,
		},
		mocks: {
			t: (app, msg) => msg,
		},
	})
}
