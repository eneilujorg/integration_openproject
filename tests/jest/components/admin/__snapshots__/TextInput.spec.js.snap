// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput is required prop should add asterik to the label text 1`] = `
<div class="text-input-label">
  some label *
</div>
`;

exports[`TextInput is required prop should not add asterik to the label text 1`] = `
<div class="text-input-label">
  some label
</div>
`;

exports[`TextInput messages should show error message if both error message and hint text are provided 1`] = `
<div class="text-input-wrapper">
  <div class="text-input">
    <div class="text-input-label">
      some label
    </div> <input id="unique-id" type="text" placeholder="" class="text-input-error">
    <div>
      <div class="text-input-error-message">
        some error message
      </div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`TextInput messages should show error message if provided 1`] = `
<div class="text-input-wrapper">
  <div class="text-input">
    <div class="text-input-label">
      some label
    </div> <input id="unique-id" type="text" placeholder="" class="text-input-error">
    <div>
      <div class="text-input-error-message">
        some error message
      </div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`TextInput messages should show hint text if provided 1`] = `
<div class="text-input-wrapper">
  <div class="text-input">
    <div class="text-input-label">
      some label
    </div> <input id="unique-id" type="text" placeholder="" class="">
    <div>
      <div class="text-input-hint">some hint message</div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`TextInput with copy button prop should render copy button if set 1`] = `
<div class="text-input-wrapper">
  <div class="text-input">
    <div class="text-input-label">
      some label
    </div> <input id="unique-id" type="text" placeholder="" class="">
    <!---->
  </div> <button disabled="disabled" title="Copy value" class="text-input-copy-value">
    <div class="text-input-icon icon-clippy"></div> <span>Copy value</span>
  </button>
</div>
`;
